### PR TITLE
refactor: replace JVM constraints with types

### PR DESCRIPTION
**Pull request by mlutze**

### CODE REVIEW

**Comment by magnus-madsen** [comment]
The point here is to increase regularity and make everything more principled, right?

**Comment by mlutze** [comment]
> The point here is to increase regularity and make everything more principled, right?
> i.e. we should have not have these extra constraints because they are equality constraints which we already support.

That's right. It especially means that the constraint solver follows the rule "I know how to unify two types if I know their kind or if I know their top-level shape." This was not the case when we had `Apply(JvmConstructor, <other stuff>)`.

**Comment by mlutze** [comment]
@magnus-madsen RFR

I'm open to suggestions for type names.

**Comment by magnus-madsen** [comment]
Can you write a very short description of how to best read the code? Thanks

File: main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala [hunk 1]
```diff
@@ -467,17 +428,23 @@ object ConstraintSolver {
...
 
-    // MRT
-    case (_, Type.Apply(Type.Cst(TypeConstructor.MethodReturnType, _), _, _)) =>
-      Result.Ok(ResolutionResult.constraints(List(TypeConstraint.Equality(tpe1, tpe2, prov)), progress = false))
-    case (Type.Apply(Type.Cst(TypeConstructor.MethodReturnType, _), _, _), _) =>
-      Result.Ok(ResolutionResult.constraints(List(TypeConstraint.Equality(tpe1, tpe2, prov)), progress = false))
+    // Java types
+    case (tpe, java@Type.JvmToType(_, _)) =>
+      for {
+        (t1, p1) <- TypeReduction.simplify(tpe, renv, loc)
```
Comment (mlutze) [inline main/src/ca/uwaterloo/flix/language/phase/typer/ConstraintSolver.scala:434]: Need to get rid of the simplify calls.

**Comment by mlutze** [comment]
> Can you write a very short description of how to best read the code? Thanks

# `Type.scala`
I introduce two new types:
- `JvmMember`
  - This is a type that corresponds to a Java field, method, or constructor. Inside it has a `JvmTemplate` which is one of
    - `JvmField(tpe, name)
    - `JvmConstructor(clazz, tpes)
    - `JvmMethod(...)`
    - `JvmStaticMethod(...)`
  - Once all the type variables are resolved, this type can be reduced to `Type.Cst(TypeConstructor.JvmField)` etc.
- `JvmToType`
  - This type extracts the return type or field type from a `Type.Cst(TypeConstructor.JvmField)` etc. when reduced.
  - This takes the place of the old `MethodReturnType` and `FieldType` types.

# `ConstraintGen.scala`

When we unify, instead of creating new constraints, we instead use equality constraints. I added the type rules in comments. They are pretty straightforward if you're familiar with the old strategy.

# `ConstraintSolver.scala`

I moved simplification into the `resolveOne` function. The logic for handling the old Jvm constraints has been moved to `TypeReduction.scala`.

I also added `JvmUnification.scala`, which only makes substitutions when one side is a variable.

# `TypeReduction.scala`

This now contains some of the logic previously in `ConstraintSolver.scala`

**Comment by magnus-madsen** [comment]
@JonathanStarup Could you help with a review?

**Comment by magnus-madsen** [comment]
Does your design support later adding a `JvmToEffect` assuming different methods have different effects?

File: main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.ast.{RigidityEnv, Type}
+
+object JvmUnification {
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala:21]: Not sure what this class is about?

**Review by magnus-madsen — COMMENTED** [review]
I think the implementation is very sensible. Good job 👍 

Picking the right representation of types is key and I think that has been achieved here.

My comments are minor and just about names.

File: main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala (hunk 2 above)
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala:21]: This is to fulfill the principle that every kind gets its own XyzUnification. Obviously not a lot happens here, but I think it's nice to delegate anyway.

**Comment by mlutze** [comment]
> Does your design support later adding a `JvmToEffect` assuming different methods have different effects?

I think that should be fine, just matching the style of `JvmToType`. There's a complication with how it interacts with Boolean unification, but that will be there in any case.

**Comment by magnus-madsen** [comment]
> > Does your design support later adding a `JvmToEffect` assuming different methods have different effects?
> 
> I think that should be fine, just matching the style of `JvmToType`. There's a complication with how it interacts with Boolean unification, but that will be there in any case.

Right-- because effects could now be learned later. Hm..

File: main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala (hunk 2 above)
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/phase/unification/JvmUnification.scala:21]: OK

File: main/src/ca/uwaterloo/flix/language/ast/Type.scala [hunk 3]
```diff
@@ -516,6 +540,97 @@ object Type {
...
+      case JvmTemplate.JvmMethod(tpe, name, tpes) => JvmTemplate.JvmMethod(f(tpe), name, tpes.map(f))
+      case JvmTemplate.JvmStaticMethod(clazz, name, tpes) => JvmTemplate.JvmStaticMethod(clazz, name, tpes.map(f))
+    }
+
+    /**
+      * Transforms `this` by executing `f` on all the types in `this`.
+      *
+      * If `f` returns `Err` for any call, this function returns `Err`.
+      */
+    def traverse[E](f: Type => Result[Type, E]): Result[JvmTemplate, E] = this match {
```
[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala:587]: OK

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala:587]: Maybe move it out of Type then?

File: main/src/ca/uwaterloo/flix/language/ast/Type.scala [hunk 4]
```diff
@@ -516,6 +540,97 @@ object Type {
...
+    * An unresolved Java type. Once the type variables are resolved, this can be reduced to a normal type.
+    */
+  case class JvmMember(template: JvmTemplate, loc: SourceLocation) extends Type with BaseType {
+    override def kind: Kind = Kind.Jvm
+  }
+
+  /**
+    * An enumeration of the unresolved Java types.
+    */
+  sealed trait JvmTemplate {
```
[the remark this answers is not in the document]
Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala:560]: `JvmMember` for this one?

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala]: I feel like JvmOp is most descriptive?

[the remark this answers is not in the document]
Comment (magnus-madsen) [reply main/src/ca/uwaterloo/flix/language/ast/Type.scala]: But I can accept JvmMember

**Comment by magnus-madsen** [comment]
Ping me when I should review again.

**Comment by mlutze** [comment]
> Ping me when I should review again.

@magnus-madsen I've addressed all the comments

File: main/src/ca/uwaterloo/flix/language/ast/Scheme.scala [hunk 5]
```diff
@@ -93,6 +93,12 @@ object Scheme {
       case Type.AssocType(sym, args, kind, _) =>
         // // Performance: Few associated types, not worth optimizing.
         Type.AssocType(sym, args.map(visitType), kind, loc)
+
+      case Type.JvmToType(tpe, loc) =>
+        Type.JvmToType(visitType(tpe), loc)
+
+      case Type.UnresolvedJvmType(template, loc) =>
+        Type.UnresolvedJvmType(template.map(visitType), loc)
```
Comment (magnus-madsen) [inline main/src/ca/uwaterloo/flix/language/ast/Scheme.scala:101]: @mlutze Need to search for "template" :)

Comment (mlutze) [reply main/src/ca/uwaterloo/flix/language/ast/Scheme.scala]: fixed!

**Review by magnus-madsen — APPROVED** [review]

**Pull request merged**